datamol/viz/_lasso_highlight.py

```diff
@@ -0,0 +1,439 @@
+# This file is thanks Christian W. Feldman
+# Christian Feldman (2021) lassohighlight [sourcecode]. https://github.com/c-feldmann/lassohighlight.
+from rdkit.Chem.Draw import rdMolDraw2D
+from rdkit.Chem.rdmolops import Get3DDistanceMatrix
+from rdkit import Chem
+from rdkit.Geometry.rdGeometry import Point2D
+import numpy as np
+from collections import defaultdict
+from collections import namedtuple
+from typing import *
+from PIL import Image
+import io
+from loguru import logger
+
+import datamol as dm
+from .utils import prepare_mol_for_drawing
+
+
+def _angle_to_coord(center: np.ndarray, angle: np.float64, radius: np.float64) -> np.ndarray:
+    """Determines a point relative to the center with distance (radius) at given angle.
+    Angles are given in rad and 0 rad correspond to north of the center point.
+    """
+
+    x = radius * np.sin(angle)
+    y = radius * np.cos(angle)
+    x += center[0]
+    y += center[1]
+    return np.array([x, y])
+
+
+def _arch_points(
+    radius: np.float64, start_ang: np.float64, end_ang: np.float64, n: int
+) -> np.ndarray:
+    """Returns an array of the shape (2, n) with equidistant points on the arch defined by
+    given radius and angles. Angles are given in rad.
+    """
+    angles = np.linspace(start_ang, end_ang, n)
+    x = radius * np.sin(angles)
+    y = radius * np.cos(angles)
+    return np.vstack([x, y]).T
+
+
+def _angle_between(center: np.ndarray, pos: np.ndarray) -> np.ndarray:
+    """Calculates the angle in rad between two points.
+    An angle of 0 corresponds to north of the center.
+    """
+    diff = pos - center
+    return np.arctan2(diff[0], diff[1])
+
+
+def _avg_bondlen(mol: Chem.Mol) -> np.ndarray:
+    """Calculates the average bond length of an rdkit.Chem.Mol object."""
+    distance_matrix = Get3DDistanceMatrix(mol)
+    bondlength_list: List = []
+    for bond in mol.GetBonds():
+        a1 = bond.GetBeginAtomIdx()
+        a2 = bond.GetEndAtomIdx()
+        bondlength_list.append(distance_matrix[a1, a2])
+    return np.mean(bondlength_list)
+
+
+Bond = namedtuple("Bond", ["angle", "neighbour_id", "bond_id"])
+
+
+class _AttachmentPointManager:
+    """AnchorManager is an invisible overlay for RDKit Atoms storing positions for arches and bond-attachment-points."""
+
+    def __init__(self, position: np.array, radius: np.float64, bond_width: np.float64):
+        self.pos = position
+        self.bond_width = bond_width
+        self.radius = radius
+        self.bonds: List[Bond] = []
+        self.bond_attachment_points: Optional[List[Tuple]] = None
```

Comment:
Below in the code, `bond_attachment_points` is init as a dict, so it seems like the type is wrong here. Either adapt it or simply remove the type.